### ISSUE TITLE
Gluon GPU memory efficiency

### ISSUE THREAD

**Issue by hnhuang** [apache/incubator-mxnet#7582]
For bugs or installation issues, please provide the following information.
The more information you provide, the more likely people will be able to help you.

## Environment info
Operating System:
Ubuntu 16.04
Package used (Python/R/Scala/Julia):
Python
MXNet version:
0.11
Python version and distribution:
2.7

I am running a vgg+fcn model on keras + tensorflow and gluon. I can set batch_size = 32 on keras + tensorflow, but I can only set batch_size = 20 on gluon. 

The last block of my model on gluon is:

# block 7
self.net = nn.Sequential(prefix='net')
self.net.add(nn.Conv3D(256, kernel_size = 3, padding = 1, activation='relu'))
self.net.add(nn.Conv3D(4096, kernel_size = 1, padding = 0, activation='relu'))
self.net.add(nn.Conv3D(4096, kernel_size = 1, padding = 0, activation='relu'))
self.net.add(nn.Conv3D(max, kernel_size = 1, activation='sigmoid'))
 
The corresponding part on keras+tf is:

x = Conv3D(256, (3, 3, 3), padding='same', activation='relu', kernel_initializer='normal', name='rpn_conv1')(base_layers)
x = Conv3D(4096, (1, 1, 1), padding='same', activation='relu', kernel_initializer='normal', name='rpn_fc1')(x)
x = Conv3D(4096, (1, 1, 1), padding='same', activation='relu', kernel_initializer='normal', name='rpn_fc2')(x)
x_dist = Conv3D(max, (1, 1, 1), activation='sigmoid', kernel_initializer='uniform', name='rpn_out_class')(x)

**Comment by hnhuang** [comment]
I also use feature pyramid network in my model. Thus there are some steps involving upsampling (or deconvolution). According to http://mxnet.io/architecture/note_memory.html, gluon should be quite GPU memory efficient.

**Comment by szha** [comment]
Do you know which commit your mxnet is on? @piiswrong recently fixed some of the memory issues in 7d6385a5 so it would be wise to first figure out whether your version is before or after that change.

**Comment by hnhuang** [comment]
Mine is commit/491f81e648639c53a68155585c53c3993a33ead5

**Comment by jermainewang** [comment]
Could you also compare the memory usage of only forward propagation? Thanks.

**Comment by piiswrong** [comment]
Could you post the full code? What card are you using? How much memory do you have?

**Comment by hnhuang** [comment]
I am using 4*TITAN Xp or TitianX. So I have 12G GPU memory in both cases for a single card. 

Which part of the code do you want to see? @piiswrong 

Let me check @jermainewang

**Comment by hnhuang** [comment]
@jermainewang For batch_size = 40, my model only use 7395MB GPU memory in forward.

**Comment by hnhuang** [comment]
If I only use vgg+fcn, then I can set batch_size = 64 for tf+keras and gluon. Here is the code:

def get_vgg16(max):
    #net = nn.HybridSequential()
    net = nn.Sequential()
    with net.name_scope():
        # Block 1
        net.add(nn.Conv3D(64, kernel_size = 3, padding = 1, activation='relu'))
        net.add(nn.Conv3D(64, kernel_size = 3, padding = 1, activation='relu'))
        net.add(nn.MaxPool3D(pool_size=(2, 2, 2), strides = (2, 2, 2)))
        # Block 2
        net.add(nn.Conv3D(128, kernel_size = 3, padding = 1, activation='relu'))
        net.add(nn.Conv3D(128, kernel_size = 3, padding = 1, activation='relu'))
        net.add(nn.MaxPool3D(pool_size=(2, 2, 2), strides = (2, 2, 2)))
        # Block 3
        net.add(nn.Conv3D(256, kernel_size = 3, padding = 1, activation='relu'))
        net.add(nn.Conv3D(256, kernel_size = 3, padding = 1, activation='relu'))
        net.add(nn.Conv3D(256, kernel_size = 3, padding = 1, activation='relu'))
        net.add(nn.MaxPool3D(pool_size=(2, 2, 2), strides = (2, 2, 2)))
        # Block 4
        net.add(nn.Conv3D(512, kernel_size = 3, padding = 1, activation='relu'))
        net.add(nn.Conv3D(512, kernel_size = 3, padding = 1, activation='relu'))
        net.add(nn.Conv3D(512, kernel_size = 3, padding = 1, activation='relu'))
        net.add(nn.MaxPool3D(pool_size=(2, 2, 2), strides = (2, 2, 2)))
        # Block 5
        net.add(nn.Conv3D(512, kernel_size = 3, padding = 1, activation='relu'))
        net.add(nn.Conv3D(512, kernel_size = 3, padding = 1, activation='relu'))
        # Block 6
        net.add(nn.Conv3DTranspose(512, kernel_size = 4, strides = (4, 4, 4), activation='relu'))
        # Block 7
        net.add(nn.Conv3D(256, kernel_size = 3, padding = 1, activation='relu'))
        net.add(nn.Conv3D(4096, kernel_size = 1, padding = 0, activation='relu'))
        net.add(nn.Conv3D(4096, kernel_size = 1, padding = 0, activation='relu'))
        net.add(nn.Conv3D(max, kernel_size = 1, activation='sigmoid'))
        return net

**Comment by hnhuang** [comment]
Here is the code for vgg+fcn+fpn:

class get_vgg16(gluon.Block):
    def __init__(self, max, **kwargs):
        super(get_vgg16, self).__init__(**kwargs)
        # c2 layer
        self.c2 = nn.Sequential(prefix='c2')
        self.c2.add(nn.Conv3D(64, kernel_size = 3, padding = 1, activation='relu'))
        self.c2.add(nn.Conv3D(64, kernel_size = 3, padding = 1, activation='relu'))
        self.c2.add(nn.MaxPool3D(pool_size=(2, 2, 2), strides = (2, 2, 2)))
        self.c2.add(nn.Conv3D(128, kernel_size = 3, padding = 1, activation='relu'))
        self.c2.add(nn.Conv3D(128, kernel_size = 3, padding = 1, activation='relu'))
        # c3 layer
        self.c3 = nn.Sequential(prefix='c3')
        self.c3.add(nn.MaxPool3D(pool_size=(2, 2, 2), strides = (2, 2, 2)))
        self.c3.add(nn.Conv3D(256, kernel_size = 3, padding = 1, activation='relu'))
        self.c3.add(nn.Conv3D(256, kernel_size = 3, padding = 1, activation='relu'))
        self.c3.add(nn.Conv3D(256, kernel_size = 3, padding = 1, activation='relu'))
        # c4 layer
        self.c4 = nn.Sequential(prefix='c4')
        self.c4.add(nn.MaxPool3D(pool_size=(2, 2, 2), strides = (2, 2, 2)))
        self.c4.add(nn.Conv3D(512, kernel_size = 3, padding = 1, activation='relu'))
        self.c4.add(nn.Conv3D(512, kernel_size = 3, padding = 1, activation='relu'))
        self.c4.add(nn.Conv3D(512, kernel_size = 3, padding = 1, activation='relu'))
        # c5 layer
        self.c5 = nn.Sequential(prefix='c5')
        self.c5.add(nn.MaxPool3D(pool_size=(2, 2, 2), strides = (2, 2, 2)))
        self.c5.add(nn.Conv3D(512, kernel_size = 3, padding = 1, activation='relu'))
        self.c5.add(nn.Conv3D(512, kernel_size = 3, padding = 1, activation='relu'))
        # u5 layer
        self.c5.add(nn.Conv3D(128, kernel_size = 1, padding = 0, activation='relu'))
        self.c5.add(nn.Conv3DTranspose(128, kernel_size = 2, strides = (2, 2, 2), activation='relu'))
        # x4 layer
        self.x4 = nn.Sequential(prefix='x4')
        self.x4.add(nn.Conv3D(128, kernel_size = 1, padding = 0, activation='relu'))
        # u4 layer
        self.u4 = nn.Sequential(prefix='u4')
        self.u4.add(nn.Conv3DTranspose(128, kernel_size = 2, strides = (2, 2, 2), activation='relu'))
        # x3 layer
        self.x3 = nn.Sequential(prefix='x3')
        self.x3.add(nn.Conv3D(128, kernel_size = 1, padding = 0, activation='relu'))
        # u3 layer
        self.u3 = nn.Sequential(prefix='u3')
        self.u3.add(nn.Conv3DTranspose(128, kernel_size = 2, strides = (2, 2, 2), activation='relu'))
        # x2 layer
        self.x2 = nn.Sequential(prefix='x2')
        self.x2.add(nn.Conv3D(128, kernel_size = 1, padding = 0, activation='relu'))
        # Block 7
        self.net = nn.Sequential(prefix='net')
        self.net.add(nn.Conv3D(256, kernel_size = 3, padding = 1, activation='relu'))
        self.net.add(nn.Conv3D(4096, kernel_size = 1, padding = 0, activation='relu'))
        self.net.add(nn.Conv3D(4096, kernel_size = 1, padding = 0, activation='relu'))
        self.net.add(nn.Conv3D(max, kernel_size = 1, activation='sigmoid'))
    def forward(self, x):
        x_c2 = self.c2(x)
        x_c3 = self.c3(x_c2)
        x_c4 = self.c4(x_c3)
        x_u5 = self.c5(x_c4)
        x4 = self.x4(x_c4) + x_u5
        x_u4 = self.u4(x4)
        x3 = self.x3(x_c3) + x_u4
        x_u3 = self.u3(x3)
        x2 = self.x2(x_c2) + x_u3
        res = self.net(x2)
        return res

**Comment by hnhuang** [comment]
I can also post my code on keras+tf if needed.

**Comment by jermainewang** [comment]
What's the memory consumption if you switch to HybridBlock? Since your code does not have any control flow and I/O manipulation, it should be fine to use HybridBlock.

**Comment by hnhuang** [comment]
Let me try.

**Comment by hnhuang** [comment]
The memory consumption is the same as before. And I don't gain any performance... Here is the code. I do call vgg_net.hybridize() in the main program.

class get_vgg16(gluon.HybridBlock):
    def __init__(self, max_radius, **kwargs):
        super(get_vgg16, self).__init__(**kwargs)
        with self.name_scope():
            # c2 layer
            self.c2 = nn.HybridSequential(prefix='c2')
            self.c2.add(nn.Conv3D(64, kernel_size = 3, padding = 1, activation='relu'))
            self.c2.add(nn.Conv3D(64, kernel_size = 3, padding = 1, activation='relu'))
            self.c2.add(nn.MaxPool3D(pool_size=(2, 2, 2), strides = (2, 2, 2)))
            self.c2.add(nn.Conv3D(128, kernel_size = 3, padding = 1, activation='relu'))
            self.c2.add(nn.Conv3D(128, kernel_size = 3, padding = 1, activation='relu'))
            # c3 layer
            self.c3 = nn.HybridSequential(prefix='c3')
            self.c3.add(nn.MaxPool3D(pool_size=(2, 2, 2), strides = (2, 2, 2)))
            self.c3.add(nn.Conv3D(256, kernel_size = 3, padding = 1, activation='relu'))
            self.c3.add(nn.Conv3D(256, kernel_size = 3, padding = 1, activation='relu'))
            self.c3.add(nn.Conv3D(256, kernel_size = 3, padding = 1, activation='relu'))
            # c4 layer
            self.c4 = nn.HybridSequential(prefix='c4')
            self.c4.add(nn.MaxPool3D(pool_size=(2, 2, 2), strides = (2, 2, 2)))
            self.c4.add(nn.Conv3D(512, kernel_size = 3, padding = 1, activation='relu'))
            self.c4.add(nn.Conv3D(512, kernel_size = 3, padding = 1, activation='relu'))
            self.c4.add(nn.Conv3D(512, kernel_size = 3, padding = 1, activation='relu'))
            # c5 layer
            self.c5 = nn.HybridSequential(prefix='c5')
            self.c5.add(nn.MaxPool3D(pool_size=(2, 2, 2), strides = (2, 2, 2)))
            self.c5.add(nn.Conv3D(512, kernel_size = 3, padding = 1, activation='relu'))
            self.c5.add(nn.Conv3D(512, kernel_size = 3, padding = 1, activation='relu'))
            # u5 layer
            self.c5.add(nn.Conv3D(128, kernel_size = 1, padding = 0, activation='relu'))
            self.c5.add(nn.Conv3DTranspose(128, kernel_size = 2, strides = (2, 2, 2), activation='relu'))
            # x4 layer
            self.x4 = nn.HybridSequential(prefix='x4')
            self.x4.add(nn.Conv3D(128, kernel_size = 1, padding = 0, activation='relu'))
            # u4 layer
            self.u4 = nn.HybridSequential(prefix='u4')
            self.u4.add(nn.Conv3DTranspose(128, kernel_size = 2, strides = (2, 2, 2), activation='relu'))
            # x3 layer
            self.x3 = nn.HybridSequential(prefix='x3')
            self.x3.add(nn.Conv3D(128, kernel_size = 1, padding = 0, activation='relu'))
            # u3 layer
            self.u3 = nn.HybridSequential(prefix='u3')
            self.u3.add(nn.Conv3DTranspose(128, kernel_size = 2, strides = (2, 2, 2), activation='relu'))
            # x2 layer
            self.x2 = nn.HybridSequential(prefix='x2')
            self.x2.add(nn.Conv3D(128, kernel_size = 1, padding = 0, activation='relu'))
            # Block 7
            self.net = nn.HybridSequential(prefix='net')
            self.net.add(nn.Conv3D(256, kernel_size = 3, padding = 1, activation='relu'))
            self.net.add(nn.Conv3D(4096, kernel_size = 1, padding = 0, activation='relu'))
            self.net.add(nn.Conv3D(4096, kernel_size = 1, padding = 0, activation='relu'))
            self.net.add(nn.Conv3D(max_radius, kernel_size = 1, activation='sigmoid'))
    def hybrid_forward(self, F, x):
        x_c2 = self.c2(x)
        x_c3 = self.c3(x_c2)
        x_c4 = self.c4(x_c3)
        x_u5 = self.c5(x_c4)
        x4 = self.x4(x_c4) + x_u5
        x_u4 = self.u4(x4)
        x3 = self.x3(x_c3) + x_u4
        x_u3 = self.u3(x3)
        x2 = self.x2(x_c2) + x_u3
        res = self.net(x2)
        return res

**Comment by jermainewang** [comment]
That's weird. Your code should perform exactly the same as writing in pure symbols. Is this how things work out for HybridBlock @piiswrong ?

**Comment by hnhuang** [comment]
I never gained any performance by using Hybrid. Even in the vgg+fcn case.

**Comment by chinakook** [comment]
The default data permutation of tensorflow is NHWC, but that of mxnet is NCHW. I think maybe the former is more memory efficient.

**Comment by hnhuang** [comment]
My observation is that if the forward computation is straight forward, then the consumption of GPU memory is almost the same (gluon vs keras+tf). But for the fully connected network, the forward computation is not straight forward, that might be the cause?

**Comment by jermainewang** [comment]
There are two possible memory issues here with regards to autograd. (1) unnecessary forward data is saved, which should be fixed by https://github.com/apache/incubator-mxnet/commit/7d6385a515dbda096560271628abf34e91e9be63 ; (2) backward memory cannot reuse forward memory since autograd requires a standalone backward graph to be executed. I would try two things:
(1) Make sure that using pure symbols give the same memory consumption as tf+keras. For example, you can try use SymbolBlock.
(2) Do forward with autograd turned on, but without calling "compute_gradient". In this case, the forward data required by backward will be saved.

If both (1) and (2) have the same memory consumption compared with tf+keras, then the reason might due to forward and backward memory cannot be reused in autograd mode.

**Comment by hnhuang** [comment]
I tried keras+mxnet. For keras+mxnet, I can set batch_size = 20, but for keras+tf, I can set batch_size = 32.

**Comment by VikingMew** [comment]
@jermainewang How to do (2)?

**Comment by szha** [comment]
@apache/mxnet-committers: This issue has been inactive for the past 90 days. It has no label and needs triage.

For general "how-to" questions, our [user forum](https://discuss.mxnet.io/) (and [Chinese version](https://discuss.gluon.ai/)) is a good place to get help.

**Comment by sxjscience** [comment]
Does anyone know the status of this issue?

**Comment by indhub** [comment]
Any update on this?

**Comment by eric-haibin-lin** [comment]
I’m comparing memory usage in symbolic mode. Will also take a look at gluon memory soon

**Comment by sandeep-krishnamurthy** [comment]
@eric-haibin-lin - Any comment on this issue and next steps? Any action items left here?

**Comment by eric-haibin-lin** [comment]
Symbolic mem usage was improved in https://github.com/apache/incubator-mxnet/pull/11142. 
I didn't look at gluon but @piiswrong made a few improvement already.

**Comment by sandeep-krishnamurthy** [comment]
Resolving in favor of broader Gluon memory optimization effort - https://github.com/apache/incubator-mxnet/issues/12226

Please reopen if closed in error.
